### ISSUE TITLE
docs(notebooks): regression masterclass — pattern 3 (Parameterized + PyroxParam + native pyrox.gp)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#77]
## Problem / Request

Port the same 7 regression models to an executed `docs/notebooks/regression_masterclass_parameterized.ipynb`, using **`pyrox._core.Parameterized` + `PyroxParam`** (declarative parameter registration with constraints + priors + autoguide selection) and, for Models 6–7, **swapping the RFF approximation for native [`pyrox.gp.GPPrior`](src/pyrox/gp/_models.py) + [`ConditionedGP`](src/pyrox/gp/_models.py)** so the GP models are no longer approximations.

This is the **third of three sibling notebooks** and the highest-abstraction one. It demonstrates the pattern that all of [`pyrox.gp._kernels`](src/pyrox/gp/_kernels.py) already uses.

## User Story

As a user who has read patterns 1 and 2, I want to see the most concise form: parameters declared as class fields with their constraints, priors, and guide types in `setup()`; the forward pass reads `self.get_param("variance")` without ever calling `numpyro.sample` directly. I also want to see Models 6 and 7 — which patterns 1 and 2 implemented as RFF *approximations* — replaced by **actual GP models** built from `pyrox.gp.GPPrior` and `ConditionedGP`, because once you accept the higher-level abstraction the workaround is unnecessary.

## Motivation

1. **Validate the `Parameterized` abstraction.** `Parameterized` already underpins every kernel in `pyrox.gp._kernels` (RBF, Matern, Periodic, Linear, RationalQuadratic, Polynomial, Cosine, White, Constant). Without a worked NN-side example, users see it as "a GP-only thing." This notebook proves it generalizes.
2. **Showcase the model/guide mode switch.** `Parameterized.set_mode("model"|"guide")` is the cleanest mechanism in pyrox for switching between prior-sampling (training) and posterior-sampling (predict) — a pattern that costs ~20 lines of plumbing in the raw form. The notebook demonstrates it once for the MLP and once for an SVI-fitted RFF.
3. **Force the abstraction-replaces-workaround moment.** RFF was invented to approximate GPs because exact GP inference was too expensive — but pyrox already ships `GPPrior` + `ConditionedGP` with conjugate Gaussian inference that's exact for Models 6–7's data scale (100 points). Patterns 1 and 2 keep RFF for pedagogy; pattern 3 *drops it* because the abstraction makes it unnecessary.
4. **Prove all-three patterns coexist.** The three notebooks landing together demonstrate that pyrox doesn't force one style — users pick the abstraction level that fits their problem.

## Proposed Structure

Same 7 model headings, but Models 6–7 are reframed:

1. **Title + Colab badge** — `# Regression Masterclass — Pattern 3: Parameterized + PyroxParam`.
2. **Background** — what `Parameterized` does: `register_param`, `set_prior`, `autoguide`, `get_param`, `set_mode`. Reference [`pyrox.gp._kernels.RBF`](src/pyrox/gp/_kernels.py#L47) as the canonical example already in production.
3. **Setup + imports + dataset** — same as patterns 1 and 2.
4. **Model 1 — Bayesian linear** — `LinearRegressor(Parameterized)`; weight registered with `dist.Normal(0, 1)` prior and `autoguide("normal")`.
5. **Model 2 — MAP MLP** — `MLP(Parameterized)`; `set_mode("model")` for MAP via `AutoDelta`.
6. **Model 3 — MC-Dropout MLP** — `MLPDropout(Parameterized)`; the dropout mask stays a `pyrox_sample` since it has no constraint to register.
7. **Model 4 — Bayesian MLP via NUTS** — Same `MLP(Parameterized)`; just change inference.
8. **Model 5 — SVR via RFF** — `SVRRegressor(Parameterized)` reusing [`RBFFourierFeatures`](src/pyrox/nn/_layers.py); only `w` is a registered param.
9. **Model 6 — Exact GP** — *Drop RFF.* Build `GPPrior(kernel=RBF(variance=1.0, lengthscale=1.0), likelihood=GaussianLikelihood(noise_var=0.05**2))`; condition on training data via `ConditionedGP`; predict on test grid. Compare against the RFF approximation from patterns 1 and 2.
10. **Model 7 — Deep GP** — Compose two `GPPrior`s using `pyrox.gp` primitives where possible; falls back to a Deep-RFF `Parameterized` module for the deep architecture not yet supported in `pyrox.gp` natively.
11. **Side-by-side comparison cell** — three-column LOC table for one model (e.g. Model 5): pattern 1 vs pattern 2 vs pattern 3.
12. **Closing markdown** — when to use which pattern; cross-link to siblings.

**No helpers.** Same constraint as patterns 1 and 2.

## Mathematical Notes

The model math is unchanged from patterns 1/2. The new conceptual material concerns the *parameter calculus* `Parameterized` exposes:

### Parameter lifecycle

For each registered parameter `θ` with init value $\theta_0$, constraint $\mathcal{C}$, and prior $p(\theta)$:

| `mode` | What `get_param("θ")` returns |
|---|---|
| `"model"` (training, with prior) | `numpyro.sample("θ", p(θ))` if a prior is set, else `numpyro.param("θ", θ₀, constraint=𝒞)` |
| `"guide"` with `autoguide="delta"` | `numpyro.param("θ", θ₀, constraint=𝒞)` — point estimate |
| `"guide"` with `autoguide="normal"` | `numpyro.sample("θ", T(Normal(loc, scale)))` where $T$ is the bijection of $\mathcal{C}$ — mean-field guide in unconstrained space |

This is the same construction `pyrox.gp` kernels use. Variance and lengthscale parameters get `constraint=positive` and `prior=LogNormal(0, 1)`, autoguide normal — and the rest follows.

### Exact GP for Models 6–7

For the conjugate Gaussian likelihood, the marginal log-likelihood and the predictive distribution are closed-form:

$$\log p(y \mid X, \theta) \;=\; -\tfrac{1}{2} y^\top (K_\theta + \sigma^2 I)^{-1} y \;-\; \tfrac{1}{2}\log\lvert K_\theta + \sigma^2 I\rvert \;-\; \tfrac{N}{2}\log(2\pi)$$

$$p(f^* \mid x^*, \mathcal{D}, \theta) \;=\; \mathcal{N}\bigl(\mu^*, \Sigma^*\bigr), \quad \mu^* = K_{*N}(K + \sigma^2 I)^{-1} y$$

For $N = 100$ this is $O(N^3) = 10^6$ operations — instant on any laptop, no RFF approximation needed. `pyrox.gp.ConditionedGP` does this exactly.

### When to drop the workaround

The RFF approximation pays off only when $D \ll N$ (i.e. $D \approx 50$ vs $N = 10^4$+). At $N = 100$ the RFF approximation **costs more** than exact GP inference (constant factors + approximation error). Pattern 3 makes the right choice: use the high-level GP primitive that already exists.

## Reference Code

`Parameterized` pattern follows the canonical `pyrox.gp` kernel form. Compare:

```python
# pyrox.gp._kernels.RBF — canonical Parameterized usage (already in the repo)
class RBF(_ParameterizedKernel):
    def setup(self):
        self.register_param("variance", jnp.array(1.0), constraint=dist.constraints.positive)
        self.register_param("lengthscale", jnp.array(1.0), constraint=dist.constraints.positive)
        self.set_prior("variance", dist.LogNormal(0.0, 1.0))
        self.set_prior("lengthscale", dist.LogNormal(0.0, 1.0))
        self.autoguide("variance", "normal")
        self.autoguide("lengthscale", "normal")

    @pyrox_method
    def __call__(self, X1, X2):
        v = self.get_param("variance")
        ell = self.get_param("lengthscale")
        return v * jnp.exp(-0.5 * jnp.sum((X1[:, None] - X2[None, :]) ** 2, -1) / ell**2)
```

Translation for Model 1 (Bayesian linear regression):

```python
from pyrox._core import Parameterized, pyrox_method

class LinearRegressor(Parameterized):
    degree: int = eqx.field(static=True)

    def setup(self):
        self.register_param("weight", jnp.zeros(self.degree + 1))
        self.set_prior("weight", dist.Normal(jnp.zeros(self.degree + 1), 1.0))
        self.autoguide("weight", "normal")

    def features(self, x):
        return jnp.stack([x ** p for p in range(self.degree + 1)], axis=-1)

    @pyrox_method
    def __call__(self, x):
        w = self.get_param("weight")
        return einsum(self.features(x), w, "n feat, feat -> n")

# Model fn is unchanged from pattern 2 (3 lines)
def model_linear(x, y=None, *, degree=1):
    net = LinearRegressor(degree=degree)
    f = numpyro.deterministic("f", net(x))
    sigma = numpyro.sample("sigma", dist.HalfNormal(1.0))
    numpyro.sample("obs", dist.Normal(f, sigma), obs=y)
```

Model 6 — exact GP, no RFF:

```python
from pyrox.gp import GPPrior, ConditionedGP, GaussianLikelihood
from pyrox.gp import RBF

prior = GPPrior(kernel=RBF(variance=1.0, lengthscale=1.0),
                likelihood=GaussianLikelihood(noise_var=0.05**2))
posterior = ConditionedGP(prior=prior, X_train=x_train[:, None], y_train=y_train)
mean, var = posterior.predict(x_test[:, None])
```

That's it for Model 6 — ~6 lines of code, exact inference, no MCMC needed when the kernel hyperparameters are fixed. Hyperparameter inference layered on top via `numpyro.infer.MCMC(NUTS(model_fn))` where `model_fn` traces through `prior.log_marginal_likelihood(x_train, y_train)`.

## References

1. [`pyrox._core.Parameterized`](src/pyrox/_core/parameterized.py) and [`PyroxParam`](src/pyrox/_core/descriptors.py) — the abstractions this notebook teaches.
2. [`pyrox.gp._kernels`](src/pyrox/gp/_kernels.py) — 9 production kernels using exactly this pattern; references in the notebook.
3. [`pyrox.gp._models.GPPrior`](src/pyrox/gp/_models.py) and [`ConditionedGP`](src/pyrox/gp/_models.py) — the native GP primitive that replaces RFF for Models 6–7.
4. Rasmussen, C. E. & Williams, C. K. I. (2006). *Gaussian Processes for Machine Learning*. MIT Press. [Ch. 2 (regression)](http://gaussianprocess.org/gpml/chapters/RW2.pdf) — the closed-form predictive used by `ConditionedGP`.
5. Tran, D., Hoffman, M. D., et al. (2018). *Simple, Distributed, and Accelerated Probabilistic Programming* (Edward2). NeurIPS. [arXiv:1811.02091](https://arxiv.org/abs/1811.02091) — closest prior art for the declarative-priors pattern.
6. Same model references as patterns 1/2: Aboleth, Rahimi & Recht 2007, Cutajar et al. 2017, Gal & Ghahramani 2016, Hoffman & Gelman 2014.
7. [`design_docs/pyrox/examples/nn/regression_masterclass_eqx.md`](design_docs/pyrox/examples/nn/regression_masterclass_eqx.md) — shared math source.

## Implementation Steps

- [ ] Author `docs/notebooks/regression_masterclass_parameterized.py` in jupytext percent format.
- [ ] Header + Colab badge + setup + imports + dataset (same skeleton as patterns 1/2).
- [ ] Models 1–5: `Parameterized` modules with `setup()` declaring constraint + prior + autoguide; `__call__` reads `get_param`.
- [ ] Model 6: build `GPPrior(kernel=RBF(...), likelihood=GaussianLikelihood(...))` + `ConditionedGP`; predict via `posterior.predict(...)`. Show comparison plot (RFF approximation from pattern 1 vs exact GP from pattern 3).
- [ ] Model 7: best-effort native composition of two `GPPrior`s; if `pyrox.gp` doesn't yet support hierarchical/deep GP composition, fall back to a `Parameterized` Deep-RFF module and **call out the gap** — that's a real follow-up issue worth filing.
- [ ] Three-column LOC comparison cell.
- [ ] Convert + execute + commit `.ipynb`.
- [ ] `mkdocs.yml` nav entry under "Tutorials → Masterclass".

## Definition of Done

- `docs/notebooks/regression_masterclass_parameterized.ipynb` is committed with embedded outputs.
- Notebook executes end-to-end under `nbconvert --execute --inplace --timeout=600`.
- Same R² assertions as pattern 1 (≥ 0.85 for non-degenerate models; ≥ 0.90 for Bayesian MLP and exact GP).
- Model 6 (exact GP) shows a tighter posterior than the RFF version from patterns 1/2 on the same data — printed as a side-by-side panel.
- Models 1–5 use `Parameterized` exclusively; no `numpyro.sample` calls inside the modules' `__call__`s (only inside model functions for `sigma` and `obs`).
- If Model 7 falls back to Deep-RFF, the notebook explicitly documents the gap and links to a new follow-up issue.

## Testing

- Same R² assertions as patterns 1 and 2.
- Additional cell asserts that `Parameterized.set_mode("guide")` flips the trace contents to `numpyro.param` sites for delta-guide leaves and to `numpyro.sample` sites for normal-guide leaves — proves the mode switch works as documented.
- Additional cell compares Model 6 exact-GP RMSE to Model 6-from-pattern-1 RFF RMSE; expect exact ≤ approximate within 1.05×.

## Documentation

- The notebook is the deliverable.
- Cross-link from [`docs/api/core.md`](docs/api/core.md) and [`docs/api/gp.md`](docs/api/gp.md) as the canonical worked example for `Parameterized` and for `GPPrior` + `ConditionedGP` end-to-end.

## Relationships

- **Parent epic:** [#36 — `[Epic] 4.C NN Uncertainty Patterns`](https://github.com/jejjohnson/pyrox/issues/36).
- **Sibling notebooks:**
  - Pattern 1 — `tree_at` raw equinox + numpyro.
  - Pattern 2 — `PyroxModule` + `pyrox_sample`.
- **Reuses (today):** [`Parameterized`](src/pyrox/_core/parameterized.py), [`PyroxParam`](src/pyrox/_core/descriptors.py), [`pyrox.gp.GPPrior`](src/pyrox/gp/_models.py), [`ConditionedGP`](src/pyrox/gp/_models.py), [`GaussianLikelihood`](src/pyrox/gp/_likelihoods.py), [`pyrox.gp.RBF`](src/pyrox/gp/_kernels.py).
- **Possible follow-up gap:** if `pyrox.gp` doesn't yet support deep / hierarchical GP composition, file an issue for `gp(deep): native deep-GP composition primitives` so Model 7 can drop its Deep-RFF fallback in a future revision.
- **Not blocked by** any other open issue. All required primitives exist today.
